# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/WSRunObject.h
# Commit: 61f680923525
# Full Hash: 61f680923525df4092faa7da9afc119ce07f4306
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-03-03 09:50:30
# Regressor Bug: 1618089
# File Overlap Count: 2
# Description:
#   Bug 1618089 - part 6: Get rid of `WSRunScanner::WSPoint` r=m_kato
#   
#   With adding new type, `EditorDOMPointInText` whose container is
#   `RefPtr<dom::Text>`, we can replace `WSRunScanner::WSPoint` and make
#   `WSRunScanner` and `WSRunObject` can use its various API.  Then, this
# ==============================================================================

diff -r e42bf0b7e1af -r 61f680923525 editor/libeditor/WSRunObject.h
--- a/editor/libeditor/WSRunObject.h	Mon Mar 02 19:02:51 2020 +0000
+++ b/editor/libeditor/WSRunObject.h	Mon Mar 02 10:08:59 2020 +0000
@@ -151,13 +151,15 @@
       : mContent(aContent), mReason(aReason) {
     AssertIfInvalidData();
   }
-  MOZ_NEVER_INLINE_DEBUG WSScanResult(nsIContent* aContent, uint32_t aOffset,
+  MOZ_NEVER_INLINE_DEBUG WSScanResult(const EditorDOMPoint& aPoint,
                                       WSType aReason)
-      : mContent(aContent), mOffset(Some(aOffset)), mReason(aReason) {
+      : mContent(aPoint.GetContainerAsContent()),
+        mOffset(Some(aPoint.Offset())),
+        mReason(aReason) {
     AssertIfInvalidData();
   }
 
-  void AssertIfInvalidData() const {
+  MOZ_NEVER_INLINE_DEBUG void AssertIfInvalidData() const {
 #ifdef DEBUG
     MOZ_ASSERT(mReason == WSType::text || mReason == WSType::normalWS ||
                mReason == WSType::br || mReason == WSType::special ||
@@ -483,21 +485,6 @@
     }
   };
 
-  // A WSPoint struct represents a unique location within the ws run.  It is
-  // always within a textnode that is one of the nodes stored in the list
-  // in the wsRunObject.  For convenience, the character at that point is also
-  // stored in the struct.
-  struct MOZ_STACK_CLASS WSPoint final {
-    RefPtr<dom::Text> mTextNode;
-    uint32_t mOffset;
-    char16_t mChar;
-
-    WSPoint() : mTextNode(nullptr), mOffset(0), mChar(0) {}
-
-    WSPoint(dom::Text* aTextNode, int32_t aOffset, char16_t aChar)
-        : mTextNode(aTextNode), mOffset(aOffset), mChar(aChar) {}
-  };
-
   /**
    * FindNearestRun() looks for a WSFragment which is closest to specified
    * direction from aPoint.
@@ -529,8 +516,10 @@
    * mTextNode is set to nullptr.
    */
   template <typename PT, typename CT>
-  WSPoint GetNextCharPoint(const EditorDOMPointBase<PT, CT>& aPoint) const;
-  WSPoint GetNextCharPointFromPointInText(const WSPoint& aPoint) const;
+  EditorDOMPointInText GetNextCharPoint(
+      const EditorDOMPointBase<PT, CT>& aPoint) const;
+  EditorDOMPointInText GetNextCharPointFromPointInText(
+      const EditorDOMPointInText& aPoint) const;
 
   /**
    * LookForNextCharPointWithinAllTextNodes() and
@@ -542,10 +531,10 @@
    * or GetPreviousCharPointFromPointInText() and returns its result.
    */
   template <typename PT, typename CT>
-  WSPoint LookForNextCharPointWithinAllTextNodes(
+  EditorDOMPointInText LookForNextCharPointWithinAllTextNodes(
       const EditorDOMPointBase<PT, CT>& aPoint) const;
   template <typename PT, typename CT>
-  WSPoint LookForPreviousCharPointWithinAllTextNodes(
+  EditorDOMPointInText LookForPreviousCharPointWithinAllTextNodes(
       const EditorDOMPointBase<PT, CT>& aPoint) const;
 
   nsresult GetWSNodes();
@@ -576,8 +565,10 @@
    * aPoint, mTextNode is set to nullptr.
    */
   template <typename PT, typename CT>
-  WSPoint GetPreviousCharPoint(const EditorDOMPointBase<PT, CT>& aPoint) const;
-  WSPoint GetPreviousCharPointFromPointInText(const WSPoint& aPoint) const;
+  EditorDOMPointInText GetPreviousCharPoint(
+      const EditorDOMPointBase<PT, CT>& aPoint) const;
+  EditorDOMPointInText GetPreviousCharPointFromPointInText(
+      const EditorDOMPointInText& aPoint) const;
 
   char16_t GetCharAt(dom::Text* aTextNode, int32_t aOffset) const;
 
@@ -786,8 +777,6 @@
   MOZ_CAN_RUN_SCRIPT nsresult AdjustWhitespace();
 
  protected:
-  using WSPoint = WSRunScanner::WSPoint;
-
   MOZ_CAN_RUN_SCRIPT nsresult PrepareToDeleteRangePriv(WSRunObject* aEndObject);
   MOZ_CAN_RUN_SCRIPT nsresult PrepareToSplitAcrossBlocksPriv();
 
@@ -807,8 +796,8 @@
    * InsertNBSPAndRemoveFollowingASCIIWhitespaces() inserts an NBSP first.
    * Then, if following characters are ASCII whitespaces, will remove them.
    */
-  MOZ_CAN_RUN_SCRIPT
-  nsresult InsertNBSPAndRemoveFollowingASCIIWhitespaces(WSPoint aPoint);
+  MOZ_CAN_RUN_SCRIPT nsresult InsertNBSPAndRemoveFollowingASCIIWhitespaces(
+      const EditorDOMPointInText& aPoint);
 
   /**
    * GetASCIIWhitespacesBounds() returns a range from start of whitespaces
